dask/array/core.py

```diff
@@ -4553,4 +4552,49 @@ def from_npy_stack(dirname, mmap_mode="r"):
     return Array(dsk, name, chunks, dtype)
 
 
+_HANDLED_FUNCTIONS = {}
+
+
+def implements(*numpy_functions):
+    """Register an __array_function__ implementation for dask.array.Array
+
+    Register that a function implements the API of a NumPy function (or several
+    NumPy functions in case of aliases) which is handled with
+    ``__array_function__``.
+
+    Parameters
+    ----------
+    \\*numpy_functions : callables
+        One or more NumPy functions that are handled by ``__array_function__``
+        and will be mapped by `implements` to a `dask.array` function.
+    """
+
+    def decorator(dask_func):
+        for numpy_function in numpy_functions:
+            _HANDLED_FUNCTIONS[numpy_function] = dask_func
+
+        return dask_func
+
+    return decorator
+
+
+def _maybe_compute(obj):
+    """Convert Array to numpy.ndarray if needed"""
+    if isinstance(obj, Array):
+        return np.array(obj)
```

Comment:
@pentschev says in https://github.com/dask/dask/pull/5032#issuecomment-507799836:
The only concern I have is with array creation here. As far as I understand, this would prevent dispatching to libraries like CuPy.

Reply:
@pentschev can you elaborate a bit? Are you worried about `obj` being a CuPy array, or a dask array with a cupy array inside? If the former, that shouldn't get to this line. A small code example would help. Then I'll add it as a test.

Reply:
It's the latter case that I'm referring to. This was one of the things that we addressed with the introduction of `_meta`, have a consistent way of not only holding non-NumPy arrays in a Dask array, but also identifying them, so we can ensure we only create CuPy arrays to do any computation with a Dask array that is composed of CuPy array(s). One way we create such arrays now is by using `np.empty_like` to get a new CuPy array and ensure that we're operating solely on CuPy arrays.

Reply:
Maybe it be better to use `dask.compute(obj)` here instead? That would work on non-NumPy arrays.

Reply:
This would work indeed. I'm kind of afraid to ask that, but any objections to using `.compute()` here @mrocklin?

Reply:
In practice, I don't think the difference between `np.asarray()` and `.compute()` is likely to make a difference -- this method is only called by numpy functions, and currently these numpy functions raise an error if they can't convert their arguments to NumPy arrays. So from a backwards compatibility perspective this would be enough.

I do think it's slightly nicer to give CuPy arrays the chance to override operations in fall-back mode, so I guess that's a reason to go for `.compute()`

Also note: in an ideal case, we would call collect all dask arrays and call `dask.compute(all_arrays)` once. This would let us take much better advantage of a parallelism. But there's a good deal more book-keeping for that version... actually, it looks like `dask.compute()` already automatically loops into arguments. So we should replace `compute_dask_args()` with a call to `dask.compute`: `args, kwargs = dask.compute(args, kwargs)`.

Reply:
I absolutely agree with all you said here @shoyer. I'm mostly asking because I'm not deeply familiarized with Dask's scheduler yet. So one thing that I imagine could happen (and maybe -- probably -- it's only in my mind) is if the scheduler may wait for operations unrelated to this array to finish before (e.g., everything's in a queue). If that case exists, it could block unnecessarily, something that wouldn't happen when calling `np.array`.

Reply:
As @shoyer says, this should probably all be replaced with a single call to `dask.compute`

Reply:
Much nicer, done.